### PR TITLE
WIP [Enhancement] Automated tests

### PR DESCRIPTION
**Pull request by truenorth**
Materialize is a marvelously popular framework, but as far as I can tell, it's lacking in the test coverage area. As people are starting to build meaningful apps on top of your good work, I believe the time has come for some improvement in this area, along with testing for cross-browser compatibility, etc...

This is a work in progress, but I'm seeking feedback as I go.

I have set up 
- [x] [testem](https://github.com/airportyh/testem) as a test runner
- [x] [qunit](https://github.com/jquery/qunit) as a test framework
- [ ] travis-ci for CI
- [ ] sauce labs for cross-browser and cross-device testing

### To try this out

`grunt testem:run:default` - should launch firefox, chrome and phantomjs to run tests

### Why this is great

Thanks to shamelessly borrowing some concepts from @rwjblue's great work on the [ember.js](https://github.com/emberjs/ember.js) testing tooling, this PR includes the ability to setup html snippets to seed test cases, and load them easily. 

For example, if you have
`/tests/pages/collapsible.html`
```html
  <ul class="collapsible" data-collapsible="accordion">
    <li>
      <div class="collapsible-header"><i class="mdi-image-filter-drama"></i>First</div>
      <div class="collapsible-body"><p>Lorem ipsum dolor sit amet.</p></div>
    </li>
    <li>
      <div class="collapsible-header"><i class="mdi-maps-place"></i>Second</div>
      <div class="collapsible-body"><p>Lorem ipsum dolor sit amet.</p></div>
    </li>
    <li>
      <div class="collapsible-header"><i class="mdi-social-whatshot"></i>Third</div>
      <div class="collapsible-body"><p>Lorem ipsum dolor sit amet.</p></div>
    </li>
  </ul>
```
You could easily load this into the DOM to setup your test module by doing the following
`tests/collapsible-test.js`
```js
module('collapsible tests', {
	setup: function () {
	  setupPage('collapsible'); // load HTML snippet
          // additional JavaScript initialization
	  $('.collapsible').collapsible({
	  	accordion : false // A setting that changes the collapsible behavior to expandable instead of the default accordion style
	  });
	}
});
```
And then you'll actually see the test DOM in the test runner

screen shot 2015-05-16 at 4 51 36 pm

Writing tests is straight forward

```js

QUnit.test('basic collapsible test', function(assert) {
	// Async mode
	QUnit.stop();

	// Starting condition
	QUnit.equal($('.collapsible li:first-child .collapsible-body').css('display'), 'block', 'First accordion is displayed initially');

	setTimeout(function() {
		QUnit.start(); // Test is done once we reach end of this scope

		// Ensure the first accordion starts collapsed
		QUnit.equal($('.collapsible li:first-child .collapsible-body').css('display'), 'none', 'First accordion is hidden after at least 350ms');

		// Click the header
		$('.collapsible li:first-child .collapsible-header').click();

		// Now the first accordion should be opened
		QUnit.equal($('.collapsible li:first-child .collapsible-header').css('display'), 'block', 'First accordion is displayed');
	}, 350);
});
```

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -479,7 +479,51 @@ module.exports = function(grunt) {
...
+            'js/pushpin.js',
+            'js/buttons.js',
+            'js/transitions.js',
+            'js/scrollFire.js',
+            'js/date_picker/picker.js',
+            'js/date_picker/picker.date.js',
+            'js/character_counter.js',
+            'tests/setup.js',
+            'tests/helpers/**/*.js',
+            'tests/**/*-test.js'
```
Comment (truenorth) [inline Gruntfile.js]: Another convention borrowed from the ember.js community - All files in the `/tests` folder that end with `-test.js` will be loaded into the test runner

**Comment by truenorth** [comment]
Before I go any deeper into this effort -- can I get some feedback from collaborators regarding whether this is something they're interested in adding to the project?

**Comment by itsmechlark** [comment]
+1

**Comment by itsmechlark** [comment]
For me, it will be great! It should be implemented. But before JS sources should be clean if you will compare with [twbs/bootstrap](https://github.com/twbs/bootstrap). It will help collaborators to fix issues in different environment.

**Comment by Dogfalo** [comment]
It is something we definitely want, we're a bit busy with work related things right now, but we'll take a closer look once things settle down. Thanks for your work.

**Comment by obsidienne** [comment]
Clearly a must.

**Comment by Dogfalo** [comment]
We'll start added some basic tests whenever we can. Thanks for you work on this @truenorth.

**Comment by truenorth** [comment]
Awesome! Happy to help add a bunch of tests on top of this, now that we have a runner and framework agreed on :)

**Comment by Dogfalo** [comment]
That would be very helpful. Also could you provide me with some guidance on setting up travis to work with what you have set up? It seems testem hangs in the travis build because it opens a webpage server up.

**Comment by omeid** [comment]
Bump.

I have wasted almost two hours now trying to find out what is wrong. So many regressions only in icon's component.

**Pull request merged**